service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/AgentDirectoryExtractor.java

```diff
@@ -0,0 +1,206 @@
+    }
+
+    private static boolean referencesPractitioner(PractitionerRole practitionerRole, List<Practitioner> practitioners) {
+        var practitionerRolePractitionerId = practitionerRole.getPractitioner().getReferenceElement().getIdPart();
+        return practitioners.stream()
+            .map(Practitioner::getIdElement)
+            .map(IdType::getIdPart)
+            .anyMatch(practitionerRolePractitionerId::equals);
+    }
+
+    private static Optional<PractitionerRole> extractRelevantPractitionerRole(Practitioner practitioner,
+        List<PractitionerRole> practitionerRoles) {
+        var practitionerId = practitioner.getIdElement().getIdPart();
+        return practitionerRoles
+            .stream()
+            .filter(practitionerRole -> practitionerRole.getPractitioner().getReferenceElement().getIdPart().equals(practitionerId))
+            .findFirst();
+    }
+
+    private static Optional<Organization> extractRelevantOrganization(PractitionerRole practitionerRole, List<Organization> organizations) {
+        var organizationId = practitionerRole.getOrganization().getReferenceElement().getIdPart();
+        return organizations
+            .stream()
+            .filter(organization -> organization.getIdElement().getIdPart().equals(organizationId))
+            .findFirst();
+    }
+
+    private static boolean containsRelevantPractitioner(Resource resource) {
+        Predicate<Resource> containsPractitioner = RESOURCE_HAS_PRACTITIONER.getOrDefault(resource.getResourceType(), (resource1) -> false);
+        return containsPractitioner.test(resource);
+    }
+
+    private static IIdType extractIIdTypes(Resource resource) {
+        Function<Resource, IIdType> extractor = RESOURCE_EXTRACT_IIDTYPE.get(resource.getResourceType());
+        return extractor.apply(resource);
+    }
+
+    private static boolean isPatientResource(Resource resource) {
+        return resource.getResourceType().equals(ResourceType.Patient);
+    }
+
+    private static boolean isNhsNumberMatching(Patient patient, String nhsNumber) {
+        return patient.getIdentifier()
+            .stream()
+            .filter(identifier -> identifier.getSystem().equals(NHS_NUMBER_SYSTEM))
+            .anyMatch(identifier -> identifier.getValue().equals(nhsNumber));
+    }
+
+    private static IIdType extractIIdTypeFromAllergyIntolerance(Resource resource) {
+        Optional<IIdType> practitionerReference = extractAsserterReference((AllergyIntolerance) resource);
+        return practitionerReference.orElseGet(() -> ((AllergyIntolerance) resource).getRecorder().getReferenceElement());
+    }
+
+    private static Optional<IIdType> extractAsserterReference(AllergyIntolerance resource) {
+        var practitionerReference = resource.getAsserter().getReferenceElement();
+        return Optional.of(practitionerReference).filter(ref -> ref.getIdPart() != null);
+    }
```

Comment:
Looks quite complex with the casting, I'm wondering if we could do something like this:
```suggestion
    private static IIdType extractIIdTypeFromAllergyIntolerance(Resource resource) {
        final Optional<AllergyIntolerance> allergyIntolerance = Optional.of(resource)
            .map(AllergyIntolerance.class::cast);

        return allergyIntolerance
            .map(AllergyIntolerance::getAsserter)
            .map(BaseReference::getReferenceElement)
            .filter(IIdType::hasIdPart)
            .orElseGet(() -> allergyIntolerance
                .map(AllergyIntolerance::getRecorder)
                .map(BaseReference::getReferenceElement).get());
    }
```